src/utils/did.rs

```diff
@@ -115,38 +294,357 @@ impl AsKnownKeyPair for Secret {
+                    Err(err_msg(
+                        ErrorKind::IllegalArgument,
+                        "Wrong codec in multibase secret material",
+                    ))?
+                }
+
+                let curve25519_point_size = 32;
+                let (d_value, x_value) = decoded_value.split_at(curve25519_point_size);
+                let base64_url_d_value = base64::encode_config(&d_value, base64::URL_SAFE_NO_PAD);
+                let base64_url_x_value = base64::encode_config(&x_value, base64::URL_SAFE_NO_PAD);
+
+                let jwk = json!({
+                    "kty": "OKP",
+                    "crv": "Ed25519",
+                    "x": base64_url_x_value,
+                    "d": base64_url_d_value
+                });
+
+                Ed25519KeyPair::from_jwk_value(&jwk)
+                    .kind(
+                        ErrorKind::Malformed,
+                        "Unable parse multibase secret material",
+                    )
+                    .map(KnownKeyPair::Ed25519)
+            }
+
             _ => Err(err_msg(
                 ErrorKind::Unsupported,
-                "Unsupported verification method type and material combination",
+                "Unsupported secret method type and material combination",
             )),
         }
     }
+}
 
-    fn key_alg(&self) -> KnownKeyAlg {
-        match (&self.type_, &self.secret_material) {
-            (SecretType::JsonWebKey2020, SecretMaterial::JWK(ref jwk)) => {
-                match (jwk["kty"].as_str(), jwk["crv"].as_str()) {
-                    (Some(kty), Some(crv)) if kty == "EC" && crv == "P-256" => KnownKeyAlg::P256,
-                    (Some(kty), Some(crv)) if kty == "EC" && crv == "secp256k1" => {
-                        KnownKeyAlg::K256
-                    }
-                    (Some(kty), Some(crv)) if kty == "OKP" && crv == "Ed25519" => {
-                        KnownKeyAlg::Ed25519
-                    }
-                    (Some(kty), Some(crv)) if kty == "OKP" && crv == "X25519" => {
-                        KnownKeyAlg::X25519
-                    }
-                    _ => KnownKeyAlg::Unsupported,
-                }
-            }
-            _ => KnownKeyAlg::Unsupported,
-        }
+#[derive(Clone, Debug, PartialEq)]
+pub enum Codec {
+    X25519Pub,
+    Ed25519Pub,
+    X25519Priv,
+    Ed25519Priv,
+}
+
+impl Codec {
+    fn codec_by_prefix(value: u32) -> Result<Codec> {
+        return match value {
+            0xEC => Ok(Codec::X25519Pub),
+            0xED => Ok(Codec::Ed25519Pub),
+            0x1302 => Ok(Codec::X25519Priv),
+            0x1300 => Ok(Codec::Ed25519Priv),
+            _ => Err(err_msg(ErrorKind::IllegalArgument, "Unsupported prefix")),
+        };
     }
 }
 
+fn _from_multicodec(value: &Vec<u8>) -> Result<(Codec, &[u8])> {
+    let mut val: Cursor<Vec<u8>> = Cursor::new(value.clone());
+    let prefix_int = val.read_unsigned_varint_32().kind(
+        ErrorKind::IllegalArgument,
```

Comment:
I believe varint's `read_unsigned_varint_32` and `write_unsigned_varint_32` throw IO Error which means that we can not read/write, and not that the prefix is invalid. So we should use `InvalidState`error kind, not `IllegalArgument`.

Reply:
fixed